### DISCUSSION TITLE
Automaticlly look for access tokens in standard places

### DISCUSSION THREAD

**Discussion by gmishkin** [PyGithub/PyGithub#3044] [Ideas]
My understanding is that the `gh` CLI will automatically check the environment variables GH_TOKEN or GITHUB_TOKEN in addition to whatever OS-specific credential store / keyring may also be available.

Anything keyring specific is probably outside the scope of this project, but what do you think about making the Github class or the token auth class support using those environment variable if a null / empty token is passed to the constructor?

**Comment by EnricoMi** [comment]
Any magic detecting the authentication still requires some user knowledge, otherwise unexpected side-effects might occur. We had some issues with `~/.netrc` files that magically (mis)-authenticated PyGithub: https://github.com/PyGithub/PyGithub/issues/1753#issuecomment-1230562207

I think having the user declare what source of authentication to use in a single line is not too bad.

The mentioned environment variables sound like a reasonable feature. That should be easy to implement as a `github.Auth.Token` sub-class, like `class EnvVarToken(Token)`, so that users can do:

    auth = Auth.EnvVarToken("GITHUB_TOKEN")
    g = Github(auth=auth)

Happy to review that feature!

**Reply by gmishkin** [reply]
Understood, what do you think about also having a no-arg constructor for the new subclass, which defaults to checking those canonical variables?

**Reply by EnricoMi** [reply]
There are so many different environments where PyGithub is being used, favouring one over the other (or supporting many of them) opens the door for complexity and risks with little benefit. Being explicit in your code about the source of authentication makes the code more readable and requires less background knowledge.

**Reply by gmishkin** [reply]
Got it. Let me just make one more nudge to see what you think and then I'll leave it alone :)

GitHub Actions itself (which may be appropriate to favor) uses this naming convention: https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication

So I'd be suggesting a constructor like

```python
class EnvVarToken:
    def __init__(self, env_var_name="GH_TOKEN")
        # ...
```

In this way, using PyGithub on GitHub Actions would require less boilerplate, similar to how when using the AWS SDK on EC2, it will pick up certain named environment variables automatically if set.

To your point though, AWS' conventions may have been around longer so there may be less code floating around using non-standard environment variable names.
